code/src/com/sixsq/nuvla/server/resources/email/sending.clj

```diff
@@ -1,29 +1,107 @@
     [clojure.java.io :as io]
     [clojure.tools.logging :as log]
     [com.sixsq.nuvla.server.resources.common.crud :as crud]
     [com.sixsq.nuvla.server.resources.configuration-nuvla :as config-nuvla]
     [com.sixsq.nuvla.server.util.response :as r]
     [java-time :as t]
+    [jsonista.core :as j]
     [postal.core :as postal]
     [selmer.parser :as tmpl]))
 
 (def base-html (slurp (io/resource "sixsq/nuvla/html-template/base.html")))
 (def trial-html (slurp (io/resource "sixsq/nuvla/html-template/trial.html")))
 (def trial-txt (slurp (io/resource "sixsq/nuvla/txt-template/trial.txt")))
 
-(defn extract-smtp-cfg
-  "Extracts the SMTP configuration from the server's configuration resource.
-   Note that this assumes a standard URL for the configuration resource."
-  [nuvla-config]
-  (when-let [{:keys [smtp-host smtp-port smtp-ssl
-                     smtp-username smtp-password]} nuvla-config]
+(def access-token-response! (atom nil))
+(def xoauth2-config! (atom nil))
+(def xoauth2-future! (atom nil))
+
+(defn post-google-refresh-access-token
+  []
+  (let [{:keys [client-id client-secret refresh-token]} @xoauth2-config!]
+    (try
+      (let [response (-> (http/post "https://accounts.google.com/o/oauth2/token"
+                                    {:headers     {"Accept" "application/json"}
+                                     :form-params {:client_id     client-id
+                                                   :client_secret client-secret
+                                                   :refresh_token refresh-token
+                                                   :grant_type    "refresh_token"}})
+                         :body
+                         j/read-value)]
+        (log/info "Refresh SMTP google access token response: "
+                  (update response "access_token" #(str (subs % 0 3) "[...]")))
+        (reset! access-token-response! response))
+      (catch Exception ex
+        (reset! access-token-response! nil)
+        (log/error "SMTP GOOGLE XOAUTH2 failed to refresh token!" (ex-data ex))))))
+
+(defn compute-next-refresh-ms
+  [access-token-response-value]
+  (-> access-token-response-value
+      (get "expires_in" 0)
+      (* 1000)
+      (- 30000)
+      (max 5000)))
+
+(defn sleep
+  [ms]
+  (^[long] Thread/sleep ms))
+
+(defn schedule-refresh-token-before-expiry
+  []
+  (reset!
+    xoauth2-future!
+    (future
+      (while true
+        (let [next-refresh-ms (compute-next-refresh-ms @access-token-response!)]
+          (log/info "SMTP xoauth2 will be automatically refreshed in:" next-refresh-ms)
+          (sleep next-refresh-ms))
+        (post-google-refresh-access-token)))))
+
+(defn refresh-token-when-no-access-token-or-on-config-change!
+  [smtp-xoauth2-config]
+  (when (or (nil? @access-token-response!)
+            (not= @xoauth2-config! smtp-xoauth2-config))
+    (reset! xoauth2-config! smtp-xoauth2-config)
+    (when @xoauth2-future! (future-cancel @xoauth2-config!))
```

Comment:
Shouldn't it be `(when @xoauth2-future! (future-cancel @xoauth2-future!))` instead? 
Otherwise, the old refresh thread may not be canceled properly.

Reply:
pushing a fix